### ISSUE TITLE
Interfaces as abstractions across Subgraphs (`@interfaceObject` and `@key` on interfaces)

### ISSUE THREAD

**Issue by pcmanus** [apollographql/federation#2277]
## Tl;dr

This issues proposes 2 new related additions:
1. a new `@interfaceObject` directive that can be applied to an object type in a subgraph as way to "abstract" what is otherwise an interface in other subgraphs.
2. support for `@key` on interface types, in order to resolve values generated by subgraph using the `@interfaceObject` feature of the previous point where needed.

The combination of those 2 features will allow a subgraph to "use" an interface transparently, without needing to know anything about the implementations of that interface (which will be provided by another subgraph).

But for more details, read on.

## Motivation

Interfaces exists to abstract, they abstract various implementations. In federation, where one the goal is to allow subgraphs to develop in relative isolation, it feels natural to want to use interface as abstractions across subgraphs. By which we mean that a subgraph should be able to use an interface as a black-box abstraction, without needing any knowledge of its implementation types.

But this is not really possible at the moment, and this issue proposes a way to achieve such abstraction.

We'll note at the offset that the proposal of this issue is meant to be a useful starting point, but that this initial proposal does not solve every interface limitations one may have with federation. We will discuss (at the end of this description) a number of possible follow-ups and show that can naturally be built on the concept introduced here.

## Running example

Let introduce a simple example to illustrate the problem first, and introduce the proposed solution next.

The example is that of an extremely simplified only shop, and consider 2 subgraphs:
1. a `products` subgraph, that defines a generic `Product` interface, and defines a number of concrete `Product` implementations (`Book`, `Movie`, etc...)
2. a `reviews` subgraph, that allows users to leave reviews on any product.

And in that example, we'll assume that reviews are generic on the actual product being reviews, meaning that we ideally want the `Review` subgraph to know the `Product` interface, but not its implementation types. Typically, if a new `Product` implementation type is added, we'd like to not have to modify the `reviews` subgraph.

Unfortunately, this is not possible today. The best one can do at the moment looks something like this:
```graphql
# Subgraph `products`
type Query {
  products: [Product!]!
}

interface Product {
  id: ID!
  description: String
  price: Float
}

type Book implements Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
  pages: Int
}

type Movie implements Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
  duration: Int
}

# ... keeping the example short, but potentially more implementations of Product ...
```
```graphql
# subgraph 'reviews'
type Query {
  bestRatedProducts(limit: Int): [Product!]!
}

interface Product {
  id: ID!
  reviews: [Review!]!
}

type Review {
  author: String
  text: String
  rating: Int
}

type Book implements Product @key(fields: "id") {
  id: ID!
  reviews: [Review!]!
}

type Movie implements Product @key(fields: "id") {
  id: ID!
  reviews: [Review!]!
}

# ... potentially more "stubs" for any other Product implementation in existence ...
```

The problem here being that the `reviews` subgraph has to define the `Product` implementation types (`Book`, `Movie`, ...), and we'll like a way to get rid of those definitions. We'll note that the reasons we need this is currently somewhat two-fold:
1. if we just remove the `Book` and `Movie` type definition, there wouldn't actually be a way to implement the `reviews` subgraph properly because of how GraphQL works. Indeed, the `bestRatedProducts` operation resolver must return objects of an object type at runtime: returning interface "values" is just not a thing in GraphQL. 
2. even if we were able to work-around that first problem, we'd still have a problem with query like
   ```graphql
   query {
     bestRatedProducts(limit: 10) {
       __typename
       price
     }
   }
   ```
   because we'd need a key to go from the `reviews` subgraph to the `products` one to get the `price`, but the keys in `products` are currently on the implementation types (or to put it another way, `products` is going to define `__resolveReference` methods "only" on `Book` and `Movie`), but if the `reviews` somehow completley abstract `Product` and don't return any information on the true concrete type, then how and where do we figure out if a given `id` returned by `bestRatedProducts` is a `Book` or a `Movie`? And btw, the `__typename` is asked, and that must return one of `Book` or `Movie` (or any other implementation type name.

Unsurprisingly, the proposal here consists of providing an answer to those 2 questions/problems.

## `@interfaceObject`: abstracting `Product` in the `reviews` subgraph.

Our 1st problem is how to allow defining the `reviews` subgraph without any mention of the `Product` implementations, but such that is can still be a valid and functional GraphQL service. And the problem is really just that having an interface with no implementations kind of doesn't work in GraphQL (to be precise, while it's rejected strictly speaking, you cannot really do anything useful with it).

As as the point is that `reviews` wants to completely abstract `Product`, and as GraphQL requires `reviews` to ultimately resolve `Product` into an object type, the proposal is that `reviews` defines `Product` as an object type directly. Of course, `Product` will continue to be an interface type if both the `products` subgraph and in the federated supergraph, so to make it explicit (to both composition and any human reader in general) that `reviews` is not making a mistake, we'll require that the `reviews` subgraph decorate its `Product` definition using the new `@interfaceObject` directive. That directive is here to say "that _object_ type is actually an _interface_ type in the federated API (but I'm abstracting it locally)".

Concretely, we will allow to define the `reviews` subgraph from our example as:
```graphql
# subgraph 'reviews'
type Query {
  bestRatedProducts(limit: Int): [Product!]!
}

type Product @key(fields: "id") @interfaceObject {
  id: ID!
  reviews: [Review!]!
}

type Review {
  author: String
  text: String
  rating: Int
}
```

This solve the question of how to define `reviews` in a way that completely abstract `Product` implementations. In particular, and to state the obvious, additions of new implementation to `Product` in `products` will require no change whatsoever to the `reviews` subgraph.

But this leads to the 2nd problem highlighted above. That is, if we have a query like:
```graphql
query {
  bestRatedProducts(limit: 10) {
    __typename
    price
  }
}
```
then we can now query the `reviews` subgraph to get the 10 best rated product `id`, but we ultimately need to be able to find out if each `id` is a `Book` or `Movie`, and as `reviews` do not tell use this, the `products` subgraph, which knows those implementations will have to.

## `@key` on interfaces

The 2nd part of this proposal is thus that `products` needs to be able to, given the `id` of a `Product`, resolve which concrete product this is. In other words, we need to support `@key` on interfaces (I'll note that the concrete meaning of a `@key` on an object type in a given subgraph amounts to declaring that the subgraph knowns how to resolve a concrete value of that type given values for the key fields; we want exactly that, but for an interface).

In other words, to complete our example, the `products` subgraph would be to modified to add a `@key(fields: "id")` on the `Product` interface, yielding (only change is the aforementioned `@key` addition):
```graphql
# Subgraph `products`
type Query {
  products: [Product!]!
}

interface Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
}

type Book implements Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
  pages: Int
}

type Movie implements Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
  duration: Int
}

# ... keeping the example short, but potentially more implementations of Product ...
```

Like for `@key` on object types, adding such `@key` on an interface will imply that a corresponding `__resolveReference` (or equivalent) will need to be defined (for the interface). The exact detail of how that resolver/function is added is subgraph library dependent (and subgraph libraries will need to be upgraded to support it; this will be "guarded" by a new version of the federation specification).

But to restate it, this function inputs will be a `Product.id` value, and it's output will need to be a concrete object, and so the primary job of that function is to determine the concrete runtime type of any product `id`.

Please note that the key resolver for interface `Product` should know how to handle all possible product `id`, the same way that a key resolver for object type `Book` should know how to handle all book `id`. That is, it's not possible for the `products` subgraph's `Book.__resolveReference` to only resolve certain books `id` but not others.

The consequence here is that putting a `@key` on an interface essentially imply that the subgraph knows of all the implementation of that interface globally. And we will add a validation that rejects a subgraph with a `@key` on an interface if that subgraph does not define all the implementations.

This conclude the initial proposal, which achieves the goal of allowing a subgraph to use an interface as a black box abstraction through what we believer are relatively simple additions to federation. 

But we do reckon that one interface related problem this does not address is the problem of "distributing" the implementations of an interface across different subgraphs. We'll now discuss potential options to that problem that build on the proposal of this issue. Those options will not be tackled in this ticket, but the point is that those option can naturally build on top of the new concept added in this ticket.

## Potential future followups

As we just mentioned, the proposal above does not provide a good way to "distribute" interface implementations across subgraphs. By "distribute", we mean a case where no single subgraph defines all the implementations of a given interface, yet federation always knows how to find the proper implementation.

To make it more concrete, the question is whether we can make an example like the following one work:
```graphql
# Subgraph 'dispatch'
type Query {
  get(id: ID!): I
}

type I @interfaceObject {
  id: ID!
  x: Int
}
```
```graphql
# Subgraph 'A'
interface I @key(fields: "id") {
  id: ID!
}

type A implements I @key(fields: "id") {
  id: ID!
  a: String
}
```
```graphql
# Subgraph 'B'
interface I @key(fields: "id") {
  id: ID!
}

type B implements I @key(fields: "id") {
  id: ID!
  b: String
}
```
In that example in particular, no subgraph knowns of both `A` and `B` (note the example could be express slightly differently by, say, removing the `dispatch` subgraph and moving the `Query` into one of the other subgraph, but this doesn't change fundamentally the problem at end).

It is hopefully clear however that (at least without making additional assumption on that example), the query planner cannot know which of subgraph `A` or `B` should be queried for a query like:
```graphql {
  get(id: "<some id>") {
    __typename
    x
    ... on A {
      a
    }
    ... on B {
      b
    }
  }
}
```

And so the only 2 general directions that seems possible are:
1. add some additional assumption so that a choice is possible. But at the end of the day, making a choice here means that `dispatch` can decide if a given `id` for `I` is a `A` or a `B`, which seems to negate the very original reason here: to not needing a subgraph to know all the implementations. However, one potentially special case here may be where one does not want to declare type `A` and `B` in `dispatch` _statically_, but is ok with the subgraph providing the concrete type _dynamically_ (at runtime).
2. or accept we cannot chose, and pessimistically request both subgraphs `A` and `B` and see which one answers (if any).

We'll discuss both of those, but let us note that each of them comes with a non-ignorable trade-off (the 1st covers only relatively specific cases, while the 2nd has potentially serious performance).

### Dynamic dispatch

The idea here is to cover specific cases where it is undesirable to add type definitions for `A` and `B` in the schema of the `dispatch` subgraph (typically because it would be inconvenient to update that schema every time a new implementation of the `I` interface is added) but were `dispatch` can nonetheless always provide the underlying implementation type (`A` or `B` here) when it resolves a `I` object at runtime. In that case, provided that the query planner has access to that information, it could dynamically dispatch to the proper subgraph based on it.

This is definitively a special case, and will only be useful in some situations. But we believe this can particular help the popular use case of wanting to provide the [Node Relay API](https://relay.dev/docs/guides/graphql-server-specification/) for a federated service (which is very similar to our example above. That API more of less assume that the `node(id: ID!)` operation can determine the type of essentially any entity, and one way to do this efficiency in a federated service can be to "encode" the entity type withing its `id`.

To make this option work, we need the `dispatch` subgraph to indicate how to provides the underlying implementation name at runtime. The simplest option would be to provide this through a specifically annotated field. And so our example above might look like (with that option):
```graphql
# Subgraph 'dispatch'
type Query {
  get(id: ID!): I
}

type I @interfaceObject {
  id: ID!
  objType: String @implementationTypename
  x: Int
}
```

### Pessimistic querying

But if one truly do not want a single subgraph to have to know all implementations (neither statically nor dynamically), then no choice can be made between the subgraphs providing implementations, and the only option will be to query all the subgraphs that can potentially return an implementation.

To be clear, it is possible to do a bit better on some very specific queries. For instance, given a query like:
```graphql
query myQuery($id: ID!) {
  get(id: $id) {
    ... on A {
      x
      a
    }
  }
}
```
then the query planner can certainly remark that only fields of type `A` are requested, and only subgraph `A` can be queried. But this is only true of fairly specific queries, and the majority of queries would have to query multiple subgraphs, and this even though only one of them would send something back.

This is still doable obviously, and there is a few different strategies that could be discussed (do we query the subgraphs in parallel? in sequence? something in between?). But no matter how you slice it, this is not very inefficient in general, and this potentially does not scale well. And so it would be particularly easy to misuse.

Overall, this is an implementable option, but it is unclear that it useful enough to be worth implementing given the performance characteristics and the potential for foot-shooting involved.

But if we do end up deciding to implement this, I think this is a behaviour that should be "opt-in". Typically, this could be done by marking the keys on `I` in both subgraph `A` and `B` as "partial" ("partial" to indicate that the subgraph only resolve a sub-set of the implementations).

**Comment by mcohen75** [comment]
At Indeed we've implemented Relay's `node` and `nodes` queries by encoding the GraphQL type into the ID. Dynamic dispatch is certainly useful here.

We've encountered other scenarios where Dynamic Dispatch would be quite useful. We've seen a pattern emerge of subgraphs that have a primary purpose of implementing search-like queries that index data across a number of other subgraphs.

Extending the example you've used here, imagine you want to implement a find query for products, and that due to team ownership or codebase size, you'd like to deliver separate services for Books and Movies.

```graphql
# Subgraph `products search`
type Query {
  findProducts(keywords: String!): [Product!]!
}

interface Product {
  id: ID!
  description: String
  price: Float
}
```

```graphql
# Subgraph `books`
type Query {
...
}

interface Product {
  id: ID!
  description: String
  price: Float
}

type Book implements Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
  pages: Int
}
```

```graphql
# Subgraph `movies`
type Query {
...
}

interface Product {
  id: ID!
  description: String
  price: Float
}

type Movie implements Product @key(fields: "id") {
  id: ID!
  description: String
  price: Float
  duration: Int
}
```

The products search subgraph in this example would have knowledge of the ids and types of data it has indexed.
